### ISSUE TITLE
spatch always looks in /usr/lib/coccinelle

### ISSUE THREAD

**Issue by jengelh** [coccinelle/coccinelle#259]
Commit bf262df5a739443af47eac197b0ce9a082ce2d67 changed:

```
 let path =
   try (Sys.getenv "COCCINELLE_HOME")
-  with Not_found->"@libdir@/coccinelle"
+  with Not_found->
+    (* was "@libdir@/coccinelle": incorrect, since "@libdir@" expands into
+       "${exec_prefix}/lib" (litteraly, without further expansion, that is
+       unlikely to be a valid path! *)
+    let exec_dir = Filename.dirname Sys.argv.(0) in
+    if Sys.file_exists (Filename.concat exec_dir "standard.iso") then
+      exec_dir
+    else
+      List.fold_left Filename.concat exec_dir
+       [Filename.parent_dir_name; "lib"; "coccinelle"]
```

As a result, even if we pass `./configure --libdir=/usr/lib64`, spatch tries to always look in /usr/lib/coccinelle instead.

The right way to do this substitution is along the lines of:

```
# Makefile.am
..._CFLAGS = -DCOCCINELLE_HOME=\"${libdir}\"

or

config.ml: config.ml.in
    sed -e 's![@]libdir@!${libdir}!' ...
```

that is, have make(!) expand it before ever passing it to the compiler.

**Comment by JuliaLawall** [comment]
On Tue, 25 May 2021, Jan Engelhardt wrote:

>
> Commit bf262df changed:
>
>  let path =
>    try (Sys.getenv "COCCINELLE_HOME")
> -  with ***@***.***@/coccinelle"
> +  with Not_found->
> +    (* was ***@***.***@/coccinelle": incorrect, since ***@***.***@" expands into
> +       "${exec_prefix}/lib" (litteraly, without further expansion, that is
> +       unlikely to be a valid path! *)
> +    let exec_dir = Filename.dirname Sys.argv.(0) in
> +    if Sys.file_exists (Filename.concat exec_dir "standard.iso") then
> +      exec_dir
> +    else
> +      List.fold_left Filename.concat exec_dir
> +       [Filename.parent_dir_name; "lib"; "coccinelle"]
>
> As a result, even if we pass ./configure --libdir=/usr/lib64, spatch tries to always look in /usr/lib/coccinelle instead.
>
> The right way to do this substitution is along the lines of:
>
> # Makefile.am
> ..._CFLAGS = -DCOCCINELLE_HOME=\"${libdir}\"
>
> or
>
> config.ml: config.ml.in
>     sed -e 's![@]libdir@!${libdir}!' ...
>
> that is, have make(!) expand it before ever passing it to the compiler.

Thanks for the suggestion.  I have forwarded it to the person who takes
care of this part of the code.

julia

**Comment by JuliaLawall** [comment]
The problem is resolved.

**Comment by elfring** [comment]
See also [the corresponding commit “Fix GH#259: Use @libdir@ for path, substituting ${exec_prefix}”](https://github.com/coccinelle/coccinelle/commit/5016596228804588c96f0059818dd937e636d7ab "Software update by Thierry Martinez from 2021-05-25").

**Comment by jacob-keller** [comment]
Hi, this appears to have broken when using a prefix such as $HOME:

warning: Can't find macro file: /home/jekeller/bin/lib/coccinelle/standard.h
warning: Can't find default iso file: /home/jekeller/bin/lib/coccinelle/standard.is
